### PR TITLE
Fix incorrect attention mask truncate in WhisperFlashAttention2

### PR DESCRIPTION
**Pull request by OliBomby**
# What does this PR do?

Fixes incorrect attention calculation when training Whisper with Flash Attention 2 and passing `decoder_attention_mask` with some values set to `False`.

This error might've been made when copying the same truncating code from the other attention implementations. The problem is that in `WhisperFlashAttention2` the dimensions have been transposed.

cc @sanchit-gandhi @ylacombe

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @eustlb

File: src/transformers/models/whisper/modeling_whisper.py [hunk 1]
```diff
@@ -421,7 +421,7 @@ def forward(
 
         causal_mask = attention_mask
         if attention_mask is not None:  # no matter the length, we just slice it
-            causal_mask = attention_mask[:, : key_states.shape[-2]]
+            causal_mask = attention_mask[:, : key_states.shape[-3]]
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/whisper/modeling_whisper.py:424]: And in the other file respectively

**Comment by vasqu** [comment]
I think this is still a pretty big bug since it highly affects the attention computations in fa2 and addresses #36585 (which has been closed under the assumption of this fix)

cc @eustlb

**Comment by greg2451** [comment]
Also affected by the issue, thanks for the fix @OliBomby. 

@eustlb would be really great to get this merged 🙏🏼

**Review by eustlb — APPROVED** [review]
LGTM! Great catch, thanks a lot 🤗
